Random hexamer bias analysis by Emma Dann
***
## Contents
* `artificial_coverage`: building and analyzing predicted coverage profiles
* `base_composition`: making profile of kmer content at specific genomic regions
* `BS_processing`: wrappers to process raw reads (trimming, mapping, meth calling) for BS converted and not converted and multiple organisms
* `deltaGprediction`: scripts for prediction of binding energies from sequencing data
* `in_silico_conversion`: scripts for in silico BS conversion of genome
* `optimization`: algorithm for primer batch optimization (implementation of Differential Evolution and E-score functions)
* `primer_concentration_effect`: Analysis of differences in coverage for different primer batches.
* `quantify_bias`: analysis of experimental coverage profiles (in genes, sites of interest etc...)
* `rounds_preamp`: analysis of effect of variable number of preamplification rounds
* `rscripts`: miscellaneous r scripts
* `snakefile`: tentative snakemake pipeline for Delta G prediction (not fully functional)
* `thrashNsnippets`: old scripts and random stuff
* `utils`: miscellaneous useful scripts (seq data handling, kmer counting, tests...)
* `hexVSprimed.py`: miscellaneous helper functions for hexamer analysis scripts (not changing location not to mess up calling)
