Random hexamer bias analysis by Emma Dann
***
## Contents
* __artificial_coverage__: building and analyzing predicted coverage profiles
* __base_composition__: making profile of kmer content at specific genomic regions
* __bs_degradation__: analysis of effect of BS induced degradation
* __coverage_peaks__: scripts for peak calling (not tested much)
* __deltaGprediction__: scripts for prediction of binding energies from sequencing data
* __in_silico_conversion__: scripts for in silico BS conversion of genome
* __mismatch_analisys__: quantification of mismatches at hexamer positions
* __optimization__: algorithm for primer batch optimization (implementation of Differential Evolution and E-score functions)
* __primer_concentration_effect__: Analysis of differences in coverage for different primer batches.
* __quantify_bias__: analysis of experimental coverage profiles (in genes, sites of interest etc...)
* __rscripts__: miscellaneous r scripts
* __snakefile__: tentative snakemake pipeline for Delta G prediction (not fully functional)
* __thrashNsnippets__: old scripts and random stuff
* __utils__: miscellaneous useful scripts (seq data handling, kmer counting, tests...)
