---
title: "Binding density primer batch"
output: html_notebook
---

```{r, eval=F}
library(parallel)
source("~/HexamerBias/rscripts/hexamer_sequence_functions.r")
```

We want to test how binding density changes for a different set of primers.

```{r}
prob.combos <- hexamerMatrix(stepSize = 0.05)
nrow(prob.combos)
```

Using `d3r.keqs` computed in `eps_and_Keq_estimation_final.Rmd`
```{r}
load("~/AvOwork/rdata/D3R_pt.RData")

d3r.pred.cov <- predict.coverage(d3r.keqs, eps.d3r)

d3r.dens <- d3r.pred.cov %>%
  mutate(binding.dens = pred.cov/abundance) 

dens <- d3r.dens %>%
  dplyr::select(template,binding.dens) 

```

Making density explanatory figure
```{r}
seq<- 'AATGCATGGATTGGTTTACGCATGCATGACTGTTCACACCTGAACAGGGTTTACGCATGGGAATACTGCAAATGCATGGATTGGTTTACGCATGCATGACTGTTCATTTACGCATGGGCTGACTGTTTTTCAGCACACAACACACCTGAACAGGGCTAGCTCGGGGTG'

l <- 20
fig.df <- data.frame(template=sapply(1:(nchar(seq)-6), function(i) substr(seq,i,i+5)) ) %>%
  inner_join(., dens) %>%
  mutate(n=1:n()) %>%
  mutate(tot.dens=rollsumr(binding.dens, k=l, fill=0)) 

# cols <- scico(length(unique(fig.df$template)), palette='berlin')
# names(cols) <- unique(fig.df$template)


p1 <- fig.df %>%  
  ggplot(., aes(n, binding.dens, fill=template)) + 
  geom_bar(stat='identity',alpha=0.7, color='black') +
  theme_classic() +
  ylab('Binding density') + xlab("Genomic coordinate") +
  guides(fill='none') +
  # geom_segment(x=which.max(fig.df$binding.dens), xend=which.max(fig.df$binding.dens)+10, y=max(fig.df$binding.dens), yend=max(fig.df$binding.dens))
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=16))

p2 <- fig.df %>%  
  ggplot(., aes(xmin=n, ymin=0, fill=template)) + 
  # geom_bar(stat='identity',alpha=0.7, width=10, position='dodge', color='black') +
  geom_rect(aes(xmax=n+l, ymax=binding.dens), alpha=0.7, color='black') +
  theme_classic() +
  ylab('Binding density') + xlab("Genomic coordinate") +
  guides(fill='none') +
  # geom_segment(x=1, xend=2, y=2, yend=3)
  geom_segment(aes(x=which.max(fig.df$binding.dens), xend=which.max(fig.df$binding.dens)+l, y=max(fig.df$binding.dens)+0.00008, yend=max(fig.df$binding.dens)+0.00008),  size=0.35) +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=16))

p3 <- fig.df %>%  
  ggplot(., aes(n, tot.dens)) + 
  geom_area(fill='grey', alpha=0.7, color='black') +
  theme_classic() +
  ylab('Predicted coverage') + xlab("Genomic coordinate") +
  guides(fill='none') +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=16)) 

p1
p2
p3

p1 + ggsave("~/AvOwork/useful_figs/dens_1_new.pdf")
p2 + ggsave("~/AvOwork/useful_figs/dens_2.pdf")
p3 + ggsave("~/AvOwork/useful_figs/dens_3.pdf")
```

### Maximize CpGs covered

To maximize the number of CpGs covered I have to enrich for hexamers that contain a CG and deplete the rest.

```{r}
hexs <- all.hexamers()
# hexs[which(grepl(hexs, pattern = "CG"))]

fc <- as.data.frame(hexs) %>%
  mutate(fc=ifelse(grepl(hexs, pattern = "CG"), 1, -1)) %>%
  rename(template=hexs)

compute.enrichment.score <- function(dens.df, fc.df){
  enrichment.score <- dens.df %>%
    inner_join(., fc.df, by="template") %>%
    mutate(dens.fc=binding.dens*fc) %>%
    summarise(score=sum(dens.fc)) %>%
    .$score
  return(enrichment.score)
  }

```

Predict coverage for different primer batches
```{r, warning=F}
test.combos <- prob.combos[which(prob.combos['pC']!=0 & prob.combos['pG']!=0),]
l.test.combos <- lapply(seq_len(nrow(test.combos)), function(i) test.combos[i,])

density.combo <- function(prob.vec, keqs.df=d3r.keqs, eps=epsilon.d3r){
  b.probs <- batch.prob.uniform(hexs=all.hexamers(), nuc.probs = prob.vec)
  pred.cov.b <- predict.coverage(keqs.df, eps, prob = b.probs)
  dens.df <- pred.cov.b %>%
    mutate(binding.dens = pred.cov/abundance)  %>% 
    dplyr::select(template, binding.dens)
  colnames(dens.df)[2] <- do.call(paste,c('dens',prob.vec, sep='_'))
  return(dens.df)
}

test <- density.combo(l.test.combos[[2]])


test.combo.density <- mclapply(sample(l.test.combos, 4), density.combo, mc.cores = 4)
func <- function(...){
  df1 = list(...)[[1]]
  df2 = list(...)[[2]]
  col1 = colnames(df1)[1]
  col2 = colnames(df2)[1]
  xxx = left_join(..., by = setNames(col2,col1))
  return(xxx)
}
dens.table <- Reduce( func, test.combo.density)

score <- unlist(lapply(test.combo.density, compute.enrichment.score, fc.df=fc))
```

Load CpG island FC hexamer enrichment
```{r}
cgi.fc <- read.csv("~/mnt/edann/hexamers/annotations_bed/CpGislands_srt_mm10.kmersFC.csv", header=F, col.names = c("template", "fc"))
plot.ranked.fc <- function(cgi.fc){
  p <- cgi.fc %>%
  arrange(-fc) %>%
  mutate(rank=min_rank(fc)) %>%
  mutate(label=ifelse(rank<=5 | rank>=max(rank)-5, as.character(template), '')) %>%
  ggplot(., aes(rank, fc, label=label)) + 
    geom_point(size=0.5, alpha=0.2) +
    geom_text_repel()
  return(p)
  }
plot.ranked.fc(cgi.fc)
compute.enrichment.score(test, cgi.fc)
```


```{r}
load('~/mnt/edann/primer_combos_density.RData')

dens.table %>% 
  select(template, dens_0_0.3_0.1_0.6) %>%
  filter(dens_0_0.3_0.1_0.6!=0) %>%
  inner_join(., cgi.fc)

dens.table %>% inner_join(select(d3r.keqs, template, abundance))
compute.enrichment.score(dens.table[1:5], cgi.fc)
```

```{r}
fc.scores <- function(dens.table, fc.df){
  dens.table.l <- map(column_to_rownames(dens.table, var='template'), function(x) data.frame(template=dens.table$template, binding.dens=x))
  scores <- map_dbl(dens.table.l, compute.enrichment.score, fc.df=fc.df)
  return(scores)
  }


dens.table.l <- map(column_to_rownames(dens.table, var='template'), function(x) data.frame(template=dens.table$template, binding.dens=x))
scores <- map_dbl(dens.table.l, compute.enrichment.score, fc.df=cgi.fc)

```

Save best hits for artificial coverage prediction
```{r}
pred.cov.table <- map(dens.table.l, function(x) x %>% 
                        inner_join(., distinct(select(d3r.keqs, template, abundance)), by='template') %>% 
                        mutate(exp=binding.dens*abundance) %>% select(template, exp))

for (name in names(pred.cov.table)) {
  df <- pred.cov.table[name][[1]] %>% full_join(select(cele.df, template)) %>% mutate(exp=ifelse(is.na(exp), 0, exp))
  write.csv(df, file= paste0("~/AvOwork/pred_cov_batches/D3R_", name, "_predcoverage.csv"), row.names = F, col.names = T, quote = F )
}

```

### How does the coverage profile of the top hits looks like?
```{r}
cov.profile.files <- list.files("~/mnt/edann/primer_conc_VAN2493/predcov_tracks", pattern = 'mat.gz', full.names = T)

high.score.profile.1 <- load.matrix("~/mnt/edann/primer_conc_VAN2493/predcov_tracks/highcov.random.42D3R_dens_0_0.3_0_0.7_predcoverage.artCov.cgis.mat.gz")
high.score.profile.2 <- load.matrix("~/mnt/edann/primer_conc_VAN2493/predcov_tracks/highcov.random.42D3R_dens_0_0.3_0.1_0.6_predcoverage.artCov.cgis.mat.gz")
low.score.profile.2 <- load.matrix("~/mnt/edann/primer_conc_VAN2493/predcov_tracks/highcov.random.42D3R_dens_0_0.1_0.7_0.2_predcoverage.artCov.cgis.mat.gz")
original.profile <- load.matrix("~/mnt/edann/primer_conc_VAN2493/predcov_tracks/D3R_original_samegenes.mat.gz")

df <- make.df.of.profiles(list(high.score1=high.score.profile.1,
                               high.score2=high.score.profile.2,
                               low.score=low.score.profile.2,
                               original.score=original.profile))

plot.genes.profile.df(df, big.labels = F, start.label = "CGI-start", "CGI-end") +
  # annotate('text',x=900, y=5.9, label=paste('PCC =', pcc), size=15) +
  ylab('Coverage (Z-score)') 
  # theme(legend.position = "bottom",
  #       plot.title = element_text(hjust = 0.5, size=33),
  #       ) 


```

```{r}
cov.track.files <- list.files("~/mnt/edann/primer_conc_VAN2493/predcov_tracks", pattern = '.bw', full.names = T)
pred.bw <- import(cov.track.files[[1]], format = 'BigWig')

```

A very ugly analysis
```{r}
f <- list.files(path='~/mnt/edann/primer_conc_VAN2493/predcov_tracks', pattern='scores.txt', full.names = T)
l.scores <- sapply(f, scan)
norm.scores <- lapply(l.scores, function(x) (x/sum(x, na.rm = T)))
names(norm.scores) <- gsub(pattern = '.+/|.scores.txt', replacement = '',names(norm.scores))
boxplot(norm.scores, outline=F, varwidth=T, las=2)

t.test(norm.scores$top.cgis, norm.scores$top.random, alternative = 'less')
t.test(norm.scores$cgis, norm.scores$random, alternative = 'less')
```

#### Improving coverage at CpG sites
Only about 1.0e+06 CpG dinucleotides are in CpG islands, so I will take all regions flanking CpG sites in the mouse genome.

1. I take the slop of 10 bps around CpG sites
```
cat hexamers/annotations_bed//cpg_sites_mm10.bed | bedtools slop -i stdin -g /hpc/hub_oudenaarden/edann/genomes/mm10/mm10.genome -b 10 > hexamers/annotations_bed/cpg_sites_mm10_10bpsSlop.bed
```

2. I compute the fold change between these and random regions using my script ```optimization/get_kmers_ROI.sh```
```{r}
cpgs.fc <- read.csv("~/mnt/edann/hexamers/annotations_bed/cpg_sites_mm10_10bpsSlop.noN.kmersFC.csv", header=F, col.names = c("template", "fc"))
plot.ranked.fc(cpgs.fc)
```


```{r}
fc.scores.cpgs <- fc.scores(dens.table, fc.df = cpgs.fc)

```


#### Improving coverage at CTCF sites

```{r}
ctcf.fc.file <- "~/mnt/edann/hexamers/strand_specific/CTCF.flank60.kmersFC.csv"
ctcf.fc <- read.csv(ctcf.fc.file, header=F, col.names = c("template", "fc"))
plot.ranked.fc(ctcf.fc)
```
```{r}
tss.fc.file <- "~/mnt/edann/hexamers/annotations_bed/TxStart_mm10.50.flank.noN.kmersFC.csv"
tss.fc <- read.csv(tss.fc.file, header=F, col.names = c("template", "fc"))

plot.ranked.fc(ctcf.fc)
plot.ranked.fc(tss.fc)
plot.ranked.fc(cgi.fc %>% filter(!grepl('N', template)))
```
```{r}
fc.scores.cgi <- fc.scores(dens.table, cgi.fc)

plot.ranked.fc(as.data.frame(fc.scores.cgi) %>% rename(fc=fc.scores.cgi) %>% rownames_to_column(var='template'))
```

```{r}
fc.scores.tss <- fc.scores(dens.table, tss.fc)
plot.ranked.fc(as.data.frame(fc.scores.tss) %>% rename(fc=fc.scores.tss) %>% rownames_to_column(var='template'))
```
```{r}
fc.scores.ctcf <- fc.scores(dens.table, fc.df = ctcf.fc)
sort(fc.scores.ctcf, decreasing = T)
plot.ranked.fc(as.data.frame(fc.scores.ctcf) %>% rename(fc=fc.scores.ctcf)%>% rownames_to_column(var='template'))
as.data.frame(fc.scores.ctcf) %>% 
  arrange(fc.scores.ctcf) %>% 
  mutate(ecdf=cume_dist(fc.scores.ctcf)) %>%
  ggplot(., aes(fc.scores.ctcf, ecdf)) +
  geom_point() +
    geom_vline(color='red', xintercept = 10)

as.data.frame(fc.scores.tss) %>% 
  arrange(fc.scores.tss) %>% 
  mutate(ecdf=cume_dist(fc.scores.tss)) %>%
  ggplot(., aes(fc.scores.tss, ecdf)) +
  geom_point() +
  geom_vline(color='red', xintercept = 10)


```

Load profiles
```{r}
profiles <- map(list.files('~/mnt/edann/pred_coverage_primer_batch_D3R/', pattern='mat.gz', full.names = T), load.matrix)
high.score.profile.1 <- load.matrix("~/mnt/edann/pred_coverage_primer_batch_D3R/random.42.mm10.noOvsD3R_dens_0_0.1_0.6_0.3_predcoverage.artCov.mat.gz")
original.profile <- load.matrix("~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42CTCF.flank60.coverage.artCov.CTCF.mat.gz")

df <- make.df.of.profiles(profiles)

plot.refpoint.profile.df(df, center = 'CTCF')
  # annotate('text',x=900, y=5.9, label=paste('PCC =', pcc), size=15) +
  ylab('Coverage (Z-score)') 
  # theme(legend.position = "bottom",
  #       plot.title = element_text(hjust = 0.5, size=33),
  #       ) 


```

#### Adding to score the weight for the binding energy
What's guiding the coverage? Is the binding energy or the primer concentration dominating?

Check the differences in densities between very similar batches
```{r}
load("~/mnt/edann/primer_combos_density_even_nreads.RData")
dens.table[1:10, 1:10]
colnames(dens.table)

dens.table$dens_0.35_0.3_0.05_0.3
dens.table %>% 
  select(dens_0.3_0.3_0.1_0.3,dens_0.35_0.3_0.05_0.3) %>%
  ggplot(., aes(log(dens_0.35_0.3_0.05_0.3), log(dens_0.3_0.3_0.1_0.3))) + geom_point()

dens.table %>% 
  select(template, dens_0.3_0.3_0.1_0.3, dens_0.4_0.15_0.05_0.4) %>%
  ggplot(., aes(log(dens_0.4_0.15_0.05_0.4), log(dens_0.3_0.3_0.1_0.3))) + geom_point()

```

```{r}
cgi.fc %>% mutate(fc=exp(fc))

```
