---
title: "DE performance"
output: html_notebook
---
```{r message = FALSE, warning=FALSE}
library(dplyr)
library(ggplot2)
library(ggrepel)
library(reshape2)
library(rtracklayer)
library(shiny)
source('~/HexamerBias/artificial_coverage/compare_peaks.r')
```
## Performance of test function
Searches for the matrix with the higher probability of having a CGCGCG primer (only one solution). Minimizes 1-p(CGCGCG)
```{r}
test.perf <- read.csv('~/mnt/edann/hexamers/strand_specific/test_function.performance.csv', header = FALSE, col.names = c('it','score'))
head(test.perf)
```

```{r}
ggplot(test.perf, aes(it,score)) +
  geom_line()
```

### Test of real function
Functions for reshaping and plotting are in DE_performance.r
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/even_cov/'
matrix.files <- list.files(opt.dir, pattern = 'matrix.csv', full.names = TRUE)
performance.mat <- read.csv(matrix.files[1], row.names = 1)
head(performance.mat)
```

```{r}
long.mat <- reshape.prob.mat(performance.mat)
head(long.mat)
max(long.mat$iter)
```

```{r}
l <- lapply(seq(1,3), function(i) plot.iteration(long.mat, i))
# png('~/AvOwork/output/DE_optimization/test_iter_matrix.png')
for(el in l){
  # png(paste0("~/AvOwork/output/DE_optimization/DE_iter-", el$data$iter[1], ".png"))
  plot(el)
  # dev.off()
}
```

### MIN RHO OF DIFFERENT TESTS

With bigger population size I get better scores and if the algorithm gets stuck in a local minimum it happens after more iterations.
```{r}
rho.files.10 <- list.files(opt.dir, pattern='_pop10_its300.DE.rho.txt', full.names = TRUE)
rho.files.20 <- list.files(opt.dir, pattern='_pop20_its300.DE.rho.txt', full.names = TRUE)
rho.files.30 <- list.files(opt.dir, pattern='_pop30_its300.DE.rho.txt', full.names = TRUE)

plot.optimization.score(rho.files.10) +
  ggtitle('Pop.size=10')

plot.optimization.score(rho.files.20) +
  ggtitle('Pop.size=20')

plot.optimization.score(rho.files.30) +
  ggtitle('Pop.size=30')

```
## Test on new optimization: maximizing density in region of interest
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/ctcf_cov/'
matrix.files <- list.files(opt.dir, pattern = '.+minimize.+matrix.csv', full.names = TRUE)
performance.mat <- read.csv("~/mnt/edann/hexamers/DEoptimization/even_cov/test1_ctcf.DE.matrix.csv", row.names = 1)
performance.mat
plot.iteration(reshape.prob.mat(performance.mat),1)
plot.iteration(reshape.prob.mat(performance.mat),99)

```
```{r}
score.ctcf.files <- list.files(opt.dir, pattern = '.+ctcf.+rho.txt', full.names = TRUE)
plot.optimization.score(score.ctcf.files) 
```
```{r}
matrix.ctcf.files <- list.files(opt.dir, pattern = '.+ctcf.+matrix', full.names = TRUE)
matrix.ctcf.files
performance.mat <- read.csv(matrix.ctcf.files[4], row.names = 1)
long.mat <- reshape.prob.mat(performance.mat)
head(performance.mat)
nrow(performance.mat)
lapply(matrix.ctcf.files, function(mat) 
  plot.iteration(reshape.prob.mat(read.csv(mat, row.names = 1)),99) +
    ggtitle(gsub(mat, pattern = '.+//', replacement = '')))
```

Test4 is so weird! Best score and ends with all Ts!!
```{r}
performance.mat <- read.csv(matrix.ctcf.files[5], row.names = 1)
long.mat <- reshape.prob.mat(performance.mat)
lapply(c(1,10,20,50,70,99),function(it) plot.iteration(long.mat, it))
```

# What does the profile full of Ts look like?
```{r}
bestmat.bw.file <- '~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test1_ctcf_bestMat.coverage.artCov.bw'
exp.bw.file <- '~/mnt/edann/hexamers/strand_specific/VAN1667_se.highcov42.bedGraph'

bestmat.bw <- import(bestmat.bw.file, format = 'BigWig')
ranges <- GRanges(seqnames = pred.bw@seqnames, 
                  ranges = IRanges(start=pred.bw@ranges@start, 
                                   end = pred.bw@ranges@start+1))
exp.bw <- import(exp.bw.file, format = 'bedGraph', which = ranges)
```
# Average coverage on CTCF sites
```{r}
bestmat.prof.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test1_ctcf_bestMat.coverage.artCov.CTCF.profile.txt', normalize = F)
bestmatTT.prof.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test6_ctcf_bestmat.coverage.artCov.CTCF.profile.txt', normalize = F)
CTCF.pred.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.CTCF.profile.txt', normalize = F)

not.norm.bestmat.df <- make.df.of.profiles(list(polyT=bestmatTT.prof.norm,
                                       best.matrix2=bestmat.prof.norm,
                                       prediction.even=CTCF.pred.norm
                                       # prediction.proportional=CTCF.pred.prop
                                       )
                                  )

bestmat.prof.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test1_ctcf_bestMat.coverage.artCov.CTCF.profile.txt')
bestmatTT.prof.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test6_ctcf_bestmat.coverage.artCov.CTCF.profile.txt')
CTCF.pred.norm <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.CTCF.profile.txt')

norm.bestmat.df <- make.df.of.profiles(list(polyT=bestmatTT.prof.norm,
                                       best.matrix2=bestmat.prof.norm,
                                       prediction.even=CTCF.pred.norm
                                       # prediction.proportional=CTCF.pred.prop
                                       )
                                  )

plot.refpoint.profile.df(not.norm.bestmat.df)
plot.refpoint.profile.df(norm.bestmat.df)
```
## Optimizing density of enriched kmers
Not total abundance

1. Popsize = 20
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/ctcf_cov/'
score.ctcf.files <- list.files(opt.dir, pattern = '.+ctcf_kmersFC.+pop20.+rho.txt', full.names = TRUE)
plot.optimization.score(score.ctcf.files) +
  theme(legend.title = element_blank())
```
2. Popsize = 30, 100 iterations
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/ctcf_cov/'
score.ctcf.files <- list.files(opt.dir, pattern = '.+ctcf_kmersFC.+pop30_its100.+rho.txt', full.names = TRUE)
plot.optimization.score(score.ctcf.files) +
  theme(legend.title = element_blank())
```
2. Popsize = 30, 100 iterations
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/ctcf_cov/'
score.ctcf.files <- list.files(opt.dir, pattern = '.+ctcf_kmersFC.+pop.+rho.txt', full.names = TRUE)
plot.optimization.score(score.ctcf.files) +
  theme(legend.title = element_blank())
```
Load coverage profiles around CTCF sites
```{r}
prof.files <- list.files(paste0(opt.dir,'predicted_cov_bestmat'), pattern = 'profile.txt', full.names = T)
profiles <- lapply(prof.files, load.profile)
names(profiles) <- gsub(prof.files, pattern = '.+500reg|.bestMat.+profile.txt', replacement = '')
df <- make.df.of.profiles(profiles)

best.scores <- sapply(score.ctcf.files, function(x) rev.default(scan(x))[1])
names(best.scores) <- gsub(names(best.scores), pattern = '.+/|.DE.+.txt', replacement = '')
df.wScores <- df %>% mutate(score=best.scores[match(df$sample, names(best.scores))])
df.wScores %>% 
  filter(sample!='test1_ctcf_kmersFC_pop30_its100') %>% 
  filter(sample!="test10_ctcf_kmersFC_pop30_its100") %>% 
  plot.refpoint.profile.df(., color = 'score') +
  guides(color = guide_colorbar(barwidth = 20, barheight = 1))
```


Not normalized 
```{r}
profiles <- lapply(prof.files, load.profile, normalize=F)
names(profiles) <- gsub(prof.files, pattern = '.+500reg|.bestMat.+profile.txt', replacement = '')
df <- make.df.of.profiles(profiles)

best.scores <- sapply(score.ctcf.files, function(x) rev.default(scan(x))[1])
names(best.scores) <- gsub(names(best.scores), pattern = '.+/|.DE.+.txt', replacement = '')
df.wScores <- df %>% mutate(score=best.scores[match(df$sample, names(best.scores))])
df.wScores %>% 
  filter(sample!='test1_ctcf_kmersFC_pop30_its100') %>% 
  filter(sample!="test10_ctcf_kmersFC_pop30_its100") %>% 
  plot.refpoint.profile.df(., color = 'score') +
  guides(color = guide_colorbar(barwidth = 20, barheight = 1))
```
## Look at the coverage profiles in a set of regions for the best matrixes
```{r}
df.wScores %>% arrange(score) %>% head(.)
pred.bw.file <- '~/mnt/edann/hexamers/DEoptimization/ctcf_cov/predicted_cov_bestmat/highcov.random.42.500regtest9_ctcf_kmersFC_pop30_its200.bestMat.coverage.artCov.bw'
pred.bw <- import(pred.bw.file, format = 'BigWig')
best.bw <- add.id(pred.bw)
norm.best.bw <- normalize.coverage(best.bw)
l.best.bw <- split(norm.best.bw, best.bw$id)
plot.expVSpred.coverage.track(l.best.bw[[2]])
```

```{r}
ctcf.bed <- '~/mnt/edann/hexamers/annotations_bed/CTCF_intestine_mm10.srt.bed'
ctcf<- import(ctcf.bed, format = 'bed')

ctcf.ovs <- findOverlaps(norm.best.bw,ctcf)
ctcf.cov <- norm.best.bw[queryHits(ctcf.ovs)]
regionsWctcf <- ctcf.cov$id
```

```{r}
lapply(sample(regionsWctcf,10), function(x) plot.cov.wAnnotation(l.best.bw[x][[1]], ctcf))
```
Compare with even primer probability predicted coverage
```{r}
even.bw <- '~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.bw'
ranges <- GRanges(seqnames = pred.bw@seqnames, 
                  ranges = IRanges(start=pred.bw@ranges@start, 
                                   end = pred.bw@ranges@start+1))
exp.bw <- import(even.bw, format = 'BigWig', which = ranges)
common.bw <- make.predVSexp.range(pred.bw, exp.bw, pred.name = 'CTCFspecific', exp.name = 'even')
norm.common.bw<- normalize.coverage(common.bw)
norm.common.bw <- add.id(norm.common.bw)
l.comon.bw <- split(norm.common.bw, norm.common.bw$id)
plot.expVSpred.coverage.track(l.comon.bw[[2]])

ctcf.ovs <- findOverlaps(norm.common.bw,ctcf)
ctcf.cov <- norm.common.bw[queryHits(ctcf.ovs)]
regionsWctcf <- ctcf.cov$id
```

```{r}
lapply(sample(regionsWctcf,10), function(x) plot.cov.wAnnotation(l.comon.bw[x][[1]], ctcf))
```
```{r}
wCTCF <- norm.common.bw[norm.common.bw$id %in% regionsWctcf]
woCTCF <- norm.common.bw[!norm.common.bw$id %in% regionsWctcf]
boxplot(wCTCF$even, wCTCF$CTCFspecific,woCTCF$even, woCTCF$CTCFspecific, outline=F, varwidth=TRUE)
```

```{r}
even.prof <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.CTCF.profile.txt')
ctcf.prof <- load.profile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test9_ctcf_kmersFC_pop30_its200.bestMat.coverage.artCov.CTCF.profile.txt')
df <- make.df.of.profiles(list(
  CTCF=ctcf.prof,
  even=even.prof
))
plot.refpoint.profile.df(df)
```
# Trying loading the whole matrix (halp)
```{r}
tab <- read.table(gzfile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test9_ctcf_kmersFC_pop30_its200.bestMat.coverage.artCov.CTCF.mat.gz'), skip = 1)
colnames(tab) <- c('chr', 'start', 'end', 'id', 'len', 'strand', seq(1,600))
tab.even <- read.table(gzfile('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.mat.gz'), skip = 1)
colnames(tab.even) <- c('chr', 'start', 'end', 'id', 'len', 'strand', seq(1,600))
```
 
```{r}
tab %>% select(-c(chr,start,end,len, strand)) %>% 
  melt(id.vars=c('id'), variable.name='position', value.name='coverage') %>%
  group_by(id) %>% mutate(zscore=(coverage-mean(coverage, na.rm=T))/sd(coverage, na.rm=T)) %>%
  ungroup() %>% 
  group_by(position) %>% 
  summarise(avg=mean(zscore, na.rm=T)) %>%
  ggplot(., aes(as.numeric(position), avg)) +
    geom_line()

tab.even %>% select(-c(chr,start,end,len, strand)) %>% 
  melt(id.vars=c('id'), variable.name='position', value.name='coverage') %>%
  group_by(id) %>% mutate(zscore=(coverage-mean(coverage, na.rm=T))/sd(coverage, na.rm=T)) %>%
  ungroup() %>% 
  group_by(position) %>% 
  summarise(avg=mean(zscore, na.rm=T)) %>%
  ggplot(., aes(as.numeric(position), avg)) +
    geom_line()
```


### EVEN COVERAGE
```{r}
opt.dir <- '~/mnt/edann/hexamers/DEoptimization/even_cov/'
score.even.files.100 <- list.files(opt.dir, pattern = '.+even_pop30_its100.+rho.txt', full.names = TRUE)
score.even.files.200 <- list.files(opt.dir, pattern = '.+even_pop30_its200.+rho.txt', full.names = TRUE)
plot.optimization.score(score.even.files.100) +
  theme(legend.title = element_blank())
plot.optimization.score(score.even.files.200) +
  theme(legend.title = element_blank())
```

# Variability in the initial population
```{r}
original.pop.good <- read.csv(paste0(opt.dir, 'test5_even_pop30_its100.DE.originalPop.csv'), row.names = 1)
original.pop.bad <- read.csv(paste0(opt.dir, 'test1_even_pop30_its100.DE.originalPop.csv'), row.names = 1)

boxplot(as.vector(cor(t(original.pop.bad))), as.vector(cor(t(original.pop.good))), ylab='correlation between vectors', names=c('bad', 'good'))
```
Doesn't seem to play a major role.

Visualizing in another way
```{r}
originalpop.even.files <- list.files(opt.dir, pattern = '.+even_pop30_its200.+originalPop.csv', full.names = TRUE)
pops <- lapply(originalpop.even.files,function(x) read.csv(x, row.names = 1) %>% reshape.prob.mat %>% mutate(p=paste0(nuc,'.',pos), name=gsub(x, pattern = '.+/|.DE.+', replacement = '')) )
mean.pops <- do.call(rbind,pops)
mean.pops %>% group_by(name,nuc) %>% summarise(avg=mean(prob)) %>%
  ggplot(., aes(nuc, avg, group=name, color=name)) +
    # facet_grid(name~.) +
    geom_line() +
    # ylim(0,1) + 
    xlab('nucleotide')
```
This is interesting: less Cs in the ones that end up having a higher score

```{r}
pred.bw.file <- '~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test2_even_pop30_its200.bestMat.coverage.artCov.bw'
exp.bw.file <- '~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.bw'

pred.bw <- import(pred.bw.file, format = 'BigWig')
ranges <- GRanges(seqnames = pred.bw@seqnames, 
                  ranges = IRanges(start=pred.bw@ranges@start, 
                                   end = pred.bw@ranges@start+1))
exp.bw <- import(exp.bw.file, format = 'BigWig', which = ranges)
common.bw <- make.predVSexp.range(pred.bw, exp.bw, pred.name = 'CTCFspecific', exp.name = 'even')
norm.common.bw<- normalize.coverage(common.bw)
norm.common.bw <- add.id(norm.common.bw)
l.comon.bw <- split(norm.common.bw, norm.common.bw$id)
```

```{r}
clean.common.bw <- l.comon.bw[which(sapply(l.common.bw, length)==4999)]
lapply(sample(clean.common.bw,10), nice.plotTrack, labels=c('Even primer conc', 'Optimized primer conc'))
```
Looks good, more noisy than not optimized data.

Profile
```{r}
opt1.prof <- load.matrix('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test2_even_pop30_its200.bestMat.coverage.artCov.mat.gz')
opt2.prof <- load.matrix('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42test5_even_pop30_its200.bestMat.coverage.artCov.mat.gz')
norm.prof <- load.matrix('~/mnt/edann/hexamers/strand_specific/artificial_coverage/highcov.random.42.artCov.mat.gz')
# exp.prof <- load.matrix('~/mnt/edann/hexamers/strand_specific/VAN1667_se.highcov42.mat.gz')

df <- make.df.of.profiles(list(
  optimized1.p=opt1.prof,
  optimized2.p=opt2.prof,
  even.p=norm.prof
))
plot.genes.profile.df(df)
ggsave('~/AvOwork/output/DE_optimization/even_optimization.pdf')

```
# Trying loading the whole matrix (halp)



