---
title: "Batch optimization"
output: html_notebook
---
Description and validation of workflow for primer batch optimization

```{r, eval=F}
library(parallel)
library(rtracklayer)
library(purrr)
library(Gviz)
library(zoo)
library(flux)
source("~/HexamerBias/deltaGprediction/binding_model_functions.r")
source("~/HexamerBias/artificial_coverage/compare_peaks.r")
source("~/HexamerBias/optimization/fc_optimization_functions.r")
```

##### Computing density for all primer batches
I generated all primer batch combinations with even nucleotide composition in all positions (with a step of 0.05 fraction) and computed coverage density based on predicted coverage for onepreamp BS-seq mouse samples (using script `run_combo_density.r`). I excluded batches with nucleotide fraction = 0 for Gs and Cs to be quicker.
```{r}
load('~/mnt/edann/primer_combos_density_onepreamp.RData')
```

##### Sequence enrichment for regions of interest 

I compute hexamer enrichment in regions of interest (log(FC) between kmer count in region of interest and random regions of the same genome) running `get_kmers_ROI.sh`.

```{r}
cgi.fc <- read.csv("~/mnt/edann/hexamers/annotations_bed/CpGislands_srt_mm10.kmersFC.csv", header=F, col.names = c("template", "fc"))
ctcf.fc <- read.csv("~/mnt/edann/hexamers/strand_specific/CTCF.flank60.kmersFC.csv", header=F, col.names = c("template", "fc"))
tss.fc <- read.csv("~/mnt/edann/hexamers/annotations_bed/TxStart_mm10.50.flank.noN.kmersFC.csv", header=F, col.names = c("template", "fc"))
cpgs.fc <- read.csv("~/mnt/edann/hexamers/annotations_bed/cpg_sites_mm10_10bpsSlop.noN.kmersFC.csv", header=F, col.names = c("template", "fc"))
```

Let's check the log(FC) distributions
```{r, echo=F}
plot.ranked.fc(cgi.fc) + ggtitle(deparse(substitute(cgi.fc)))
plot.ranked.fc(ctcf.fc) + ggtitle(deparse(substitute(ctcf.fc)))
plot.ranked.fc(tss.fc) + ggtitle(deparse(substitute(tss.fc)))
plot.ranked.fc(cpgs.fc) + ggtitle(deparse(substitute(cpgs.fc)))
```

##### Compute enrichment score for batches
For each possible primer batch I compute an "enrichment score" for the region of interest, multiplying the predicted hexamer density by the fold change enrichment for that hexamer and summing up all the resulting scores.

```{r}
cgi.score <- fc.scores(dens.table, fc.df = cgi.fc)
tss.score <- fc.scores(dens.table, fc.df = tss.fc)
ctcf.score <- fc.scores(dens.table, fc.df = ctcf.fc)
cpgs.score <- fc.scores(dens.table, fc.df = cpgs.fc)
```
```{r, warning=F, echo=F}
as.data.frame(cgi.score) %>% rename(score=cgi.score) %>% rownames_to_column(var='template') %>%
  plot.ranked.score() +
  ggtitle('CGI')

as.data.frame(ctcf.score) %>% rename(score=ctcf.score) %>% rownames_to_column(var='template') %>%
  plot.ranked.score() +
  ggtitle('CTCF')

as.data.frame(tss.score) %>% rename(score=tss.score) %>% rownames_to_column(var='template') %>%
  plot.ranked.score() +
  ggtitle('TSS')

```
Note how the best are always the same, indicating that we are enriching for CG sites in general, and that for the CpG islands we get higher top scores (> 40) than the rest. 

```{r}
ctcf.score.df <- as.data.frame(ctcf.score) %>% 
  rename(score=ctcf.score) %>% 
  rownames_to_column(var='template') %>%
  mutate(ROI='CTCF sites')
cgi.score.df <- as.data.frame(cgi.score) %>% 
  rename(score=cgi.score) %>% 
  rownames_to_column(var='template') %>%
  mutate(ROI='CpG islands')
tss.score.df <- as.data.frame(tss.score) %>% 
  rename(score=tss.score) %>% 
  rownames_to_column(var='template') %>%
  mutate(ROI='TSS')
cpgs.score.df <- as.data.frame(cpgs.score) %>% 
  rename(score=cpgs.score) %>% 
  rownames_to_column(var='template') %>%
  mutate(ROI='CpG sites')

full.rank.pl <- list(ctcf.score.df, cgi.score.df, tss.score.df) %>%
  reduce(bind_rows) %>%
  group_by(ROI) %>%
  arrange(score) %>%
  mutate(rank=min_rank(score)) %>%
  ungroup() %>%
  plot.ranked.score(massage=F) +
  facet_grid(.~ROI) +
  theme(strip.text=element_text(size=15)) +
  xlab('Batch rank') + ylab('E-score')


full.rank.pl + ggsave("~/AvOwork/output/DE_optimization/rank_score_full.pdf", width=15)
```

###### Are the top e-scores driving the final score?
```{r}

sort(cgi.score, decreasing = T) %>% head
select(dens.table, template, dens_0_0.2_0_0.8) %>% 
  rename(binding.dens=dens_0_0.2_0_0.8) %>%
  inner_join(., cgi.fc, by="template") %>%
  mutate(dens.fc=binding.dens*fc) %>%
  mutate(cs=cume_dist(dens.fc)) %>%
  # filter(dens.fc!=0) %>%
  ggplot(aes(dens.fc, cs)) +
    geom_point(alpha=0.3, size=0.5)
  geom_histogram(bins=1000)   
geom_vline(xintercept = cgi.score['dens_0_0.2_0_0.8'])

select(dens.table, template, one_of(c(bottom.scores, top.scores, 'dens_0.25_0.25_0.25_0.25'))) %>%
  melt(id.vars='template', value.name='binding.dens', variable.name='batch') %>%
  # rename(binding.dens=dens_0.25_0.25_0.25_0.25) %>%
  inner_join(., cgi.fc, by="template") %>%
  inner_join(., as.data.frame(tss.score) %>% rownames_to_column(var='batch') , by='batch') %>%
  mutate(dens.fc=binding.dens*fc) %>%
  group_by(batch) %>%
  mutate(cs=cume_dist(dens.fc)) %>%
  ungroup() %>%
  mutate(batch=substr(batch, 1,23)) %>%
  # filter(dens.fc!=0) %>%
  ggplot(aes(dens.fc, cs, color=tss.score)) +
    geom_point(alpha=0.5, size=0.5) +
  facet_wrap(~batch, scales = 'free_x', ncol=5) +
  # ggtitle(label=batch.name) +
  xlab('e-score') + ylab('cumulative freq.') +
  scale_color_gradient2(midpoint = 10, name='E-score') +
  theme(axis.title = element_text(size=20),
        axis.text = element_text(size=16),
        strip.text = element_text(size=12),
        legend.text = element_text(size=16),
        legend.title = element_text(size=18)) +
  ggsave("~/AvOwork/formatted_figs/suppl_figs_final/cum_dist_escores_tss.pdf", height=15, width=15)

  NULL
#   geom_histogram(bins=1000)   
# geom_vline(xintercept = cgi.score['dens_0_0.2_0_0.8'])

bottom.scores
```

##### Make predicted coverage tracks for all 
Save and do predicted track with script `strand_specific_artificial_coverage.py`

```{r, warning=F}
bottom.scores <- cgi.score %>% sort %>% head(10) %>% names
top.scores <- cgi.score %>% sort(decreasing = T) %>% head(10) %>% names

load("~/mnt/edann/VAN2591/mm10.onepreamp.keqs.RData")
dens.table.l <- map(column_to_rownames(dens.table, var='template'), function(x) data.frame(template=dens.table$template, binding.dens=x))
suppressWarnings(
  pred.cov.table <- map(dens.table.l, function(x) x %>% 
                        inner_join(., distinct(select(mm10.onepreamp.keqs, template, abundance)), by='template') %>% 
                        mutate(exp=binding.dens*abundance) %>% select(template, exp))
  )

for (name in c(bottom.scores,top.scores, 'dens_0.25_0.25_0.25_0.25')) {
  df <- pred.cov.table[name][[1]] %>% 
    full_join(data.frame(template=all.hexamers()), by='template') %>%
    mutate(exp=ifelse(is.na(exp), 0, exp))
  write.csv(df, file= paste0("~/AvOwork/pred_cov_batches/mm10_onepreamp_", name, "_predcoverage.csv"), row.names = F, col.names = T, quote = F )
}
```

##### Compute density yield in regions of interest 
To validate the optimization, I compute the yield for a batch in a set of regions of interest as: 
$$
Y_b(ROI)=\frac{AUC(c(p)_b + a)_{in}}{AUC(c(p)_b + a)_{out}}
$$
where $AUC$ is Area Under the Curve, $c(p)$ is the predicted coverage track, $a$ is an integer used for scaling everything to positive values (here $a=5$) (implemented in `compute_yield_track_dir.r`). 
```{r}
cgis.yield <- read.csv("~/mnt/edann/pred_coverage_primer_batch_mm10_onepreamp/yield_CpGislands_srt_mm10.csv")
ctcf.yield <- read.csv("~/mnt/edann/pred_coverage_primer_batch_mm10_onepreamp/yield_CTCF.flank60.csv")
tss.yield <- read.csv("~/mnt/edann/pred_coverage_primer_batch_mm10_onepreamp/yield_TxStart_mm10.50.flank.csv")
```

Plot yield, tracks are sorted by $E$ score.
```{r}
massage.yield <- function(yield.df, fc.score, relative.to=NULL){
    df <- yield.df %>%
    mutate(track=gsub(track, pattern = name.pattern, replacement = '')) 
    # mutate(delta_yield=yield-1.3973866) %>%
  if (!is.null(relative.to)) {
    df <- df %>% mutate(yield = yield - filter(., track==relative.to)$yield) 
  }
  df <- df %>%
    mutate(fc.score=fc.score[match(track, names(fc.score))]) %>%
    arrange(fc.score) %>%
    mutate(track=factor(track, levels = track))
  return(df)
}

plot.yield.roi <- function(yield.score.df, name.pattern='random.+onepreamp_|_predcove.+', score.midpoint=20, relative.to=NULL){
  pl <- yield.score.df %>%
    ggplot(., aes(x=track, yield, fill=fc.score)) +
    geom_bar(stat='identity') +
    coord_flip() +
    scale_fill_gradient2(midpoint = score.midpoint) +
    theme_minimal()
  if (!is.null(relative.to)) {
    pl <- pl + ylab(paste('yield (relative to', relative.to, ')')) 
  }
  return(pl)
  }
```

```{r, echo=F}
massage.yield(cgis.yield, fc.score = cgi.score) %>% plot.yield.roi() + ggtitle('CpG islands')
massage.yield(ctcf.yield, fc.score = ctcf.score) %>% plot.yield.roi() + ggtitle('CTCF sites')
massage.yield(tss.yield, fc.score = tss.score) %>% plot.yield.roi() + ggtitle('Transcription start sites')
```



###### Yield relative to random batch
```{r, echo=F}
massage.yield(cgis.yield, fc.score = cgi.score, relative.to = 'dens_0.25_0.25_0.25_0.25') %>% plot.yield.roi(relative.to = 'dens_0.25_0.25_0.25_0.25') + ggtitle('CpG islands')
massage.yield(ctcf.yield, fc.score = ctcf.score, relative.to = 'dens_0.25_0.25_0.25_0.25') %>% plot.yield.roi(relative.to = 'dens_0.25_0.25_0.25_0.25') + ggtitle('CTCF sites')
massage.yield(tss.yield, fc.score = tss.score, relative.to = 'dens_0.25_0.25_0.25_0.25') %>% plot.yield.roi(relative.to = 'dens_0.25_0.25_0.25_0.25') + ggtitle('Transcription start sites')
```

The highest difference in yield I find for the CpG islands (where I also had higher top scores), probably indicating that sequence composition inside CGIs is more specific, and thus "enrichable". 

Facet everything.
```{r}
cgis.yield %>%
  full_join(., tss.yield, by='track') %>%
  full_join(., ctcf.yield, by='track') 

faceted.yield.delta <- list(cgis.yield, tss.yield, ctcf.yield) %>%
  map(function(x) mutate(x, track=gsub(track, pattern = name.pattern, replacement = ''))) %>%
  map2(list(cgi.score, tss.score, ctcf.score), function(x,y) mutate(x, fc.score=y[match(track, names(y))])) %>%
  reduce(bind_rows) %>%
  group_by(ROI) %>%
  mutate(yield.delta = yield - yield[track=='dens_0.25_0.25_0.25_0.25'],
         fc.score.delta = fc.score - fc.score[track=='dens_0.25_0.25_0.25_0.25']) %>%
  arrange(fc.score) %>%
  mutate(track=factor(track, levels = track)) %>%
  ungroup() 

faceted.yield.delta.plot <- faceted.yield.delta %>%
  # head
  ggplot(., aes(x=track, y=yield.delta, fill=fc.score)) +
    facet_grid(.~ROI) +
    geom_bar(stat='identity') +
    coord_flip() +
    scale_fill_gradient2(name='E-score') +
    theme_bw() +
    theme(strip.text = element_blank(),
          legend.position = 'bottom',
          legend.title = element_text(size=20),
          legend.text = element_text(size=14),
          axis.title = element_text(size=16),
          axis.text.x = element_text(size=14))

faceted.yield.delta %>%
  # head
  ggplot(., aes(x=fc.score, y=yield, color=ROI)) +
    # facet_grid(.~ROI) +
  geom_point(alpha=0.8) +
  stat_cor()

    geom_bar(stat='identity') +
    coord_flip() +
    scale_fill_gradient2(name='E-score') +
    theme_bw() +
    theme(strip.text = element_blank(),
          legend.position = 'bottom',
          legend.title = element_text(size=20),
          legend.text = element_text(size=14),
          axis.title = element_text(size=16),
          axis.text.x = element_text(size=14))

```

##### Making plot to label batches
```{r}
batch.df <- map(strsplit(names(dens.table)[-1],  '_'), function(x) as.numeric(x[2:5])) %>% reduce(rbind) %>% data.frame(row.names = NULL)
colnames(batch.df) <- c('A', 'C', 'T', 'G')
df <- data.frame(batch=names(dens.table)[-1]) %>%
  bind_cols(., batch.df)

plot.batch.composition <- function(batch.df, fc.score){
  pl1 <- batch.df %>%
    melt %>%
    filter(batch %in% c(bottom.scores,top.scores, 'dens_0.25_0.25_0.25_0.25')) %>%
    mutate(fc.score=fc.score[match(batch, names(fc.score))]) %>%
    arrange(fc.score) %>%
    # head
    mutate(batch=factor(batch, levels = rev(unique(batch)))) %>%
    ggplot(., aes(batch, variable,label=variable, size=value, color=variable)) + 
    facet_grid(batch~.) +
    theme_bw() +
    geom_text(x=0.5) +
    coord_fixed(ratio=1/2) +
    scale_size(range=c(0,10)) +
    guides(color='none') +
    coord_flip() +
    theme(
      strip.background = element_blank(),
      strip.text.y = element_blank(), 
      panel.grid.major = element_blank(),
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks = element_blank(),
      legend.position = 'bottom'    
    )
  return(pl1)
}

grid.batch.yield <- function(yield.df, fc.score, main.title){
  pl2 <- plot.yield.roi(yield.df, fc.score = fc.score, relative.to = 'dens_0.25_0.25_0.25_0.25',score.midpoint = 15) + 
  ggtitle(main.title) +
  theme(axis.text.y = element_blank(),
        axis.title.y = element_blank())
  pl1 <- plot.batch.composition(df, fc.score)
  gr <- grid.arrange(pl1, pl2, ncol=2, widths=0.5:2, heights=3)
  return(gr)
  }

grid.ctcf <- grid.batch.yield(ctcf.yield, ctcf.score, main.title='CTCF sites') 
ggsave(grid.ctcf, filename = '~/AvOwork/output/DE_optimization/yield_ctcf_grid.pdf')
grid.cgi <- grid.batch.yield(cgis.yield, cgi.score,  main.title='CpG islands')
ggsave(grid.cgi, filename = '~/AvOwork/output/DE_optimization/yield_cgi_grid.pdf')
grid.tss <- grid.batch.yield(tss.yield, tss.score, main.title='Transcription Start Sites')
ggsave(grid.tss, filename = '~/AvOwork/output/DE_optimization/yield_tss_grid.pdf')
```

With faceted plot 
```{r}
pl1 <- plot.batch.composition(df, cgi.score)
pl1
pl2 <- faceted.yield.delta.plot +
    theme(axis.text.y = element_blank(),
        axis.title.y = element_blank())
  
gr <- grid.arrange(pl1, pl2, ncol=2, widths=0.5:2, heights=3)
ggsave(gr, filename = '~/AvOwork/output/DE_optimization/yield_grid_all.pdf', width=12, height=9)
```



### Making scheme to describe algorithm
```{r}
dens <- filter(dens.table[,10], dens.table$template %in% smp.fc$template) %>%
  rename(density=dens_0.4_0.05_0_0.55) %>%
  mutate(template=1:n()) 
dens.plot <- dens %>%
  ggplot(aes(template,y=0, yend=density, xend=template)) +
    geom_segment(size=1) +
    geom_point(aes(y=density), size=1) +
  theme_bw() +
  coord_flip() +
    ylab('coverage density') +
  geom_hline(yintercept = 0) +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=18))

smp.fc <- cgi.fc %>% 
  sample_n(20) 
smp.fc.plot <- smp.fc %>%  
  mutate(template=1:n()) %>%
  ggplot(aes(template,y=0, yend=fc, xend=template)) +
    geom_segment(size=1) +
    geom_point(aes(y=fc), size=1) +
  theme_bw() +
  coord_flip() +
    ylab('log(FC)') +
  geom_hline(yintercept = 0) +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=18))

e.score.plot <- dens %>%
  bind_cols(smp.fc) %>%
  mutate(e.score=fc*density) %>%
  ggplot(aes(template,y=0, yend=e.score, xend=template)) +
    geom_segment(size=1) +
    geom_point(aes(y=e.score), size=1) +
  theme_bw() +
  coord_flip() +
  ylab('e-score') +
  geom_hline(yintercept = 0) +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(size=18))

dens.plot + ggsave('~/AvOwork/useful_figs/fc_algorithm_scheme_dens.pdf')
smp.fc.plot + ggsave('~/AvOwork/useful_figs/fc_algorithm_scheme_fc.pdf')
e.score.plot + ggsave('~/AvOwork/useful_figs/fc_algorithm_scheme_escore.pdf')
```


##### Making E-score table for report
```{r}
d <- data.frame(cgi.score, tss.score, ctcf.score) %>% 
  rownames_to_column(var='batch') %>%
  arrange(-cgi.score) %>%
  top_n(n=10) 

split.batch.name <- function(batch.name){
  spl.name <- as.numeric(strsplit(batch.name, split = '_')[[1]][-1])*100
  names(spl.name) <- c('A', 'C', 'T', 'G')
  return(spl.name)
  }

formatted.d <- d %>% 
  mutate(A=sapply(batch, function(b) split.batch.name(b)['A']),
         C=sapply(batch, function(b) split.batch.name(b)['C']),
         T=sapply(batch, function(b) split.batch.name(b)['T']),
         G=sapply(batch, function(b) split.batch.name(b)['G'])) %>%
  # mutate(rank=1:n()) %>%
  select( A,C,T,G, tss.score, cgi.score, ctcf.score)

xtable(formatted.d)
```