---
title: "Different template concentration analysis"
output: html_notebook
---

Load some functions from```compareHexamerUsage```

```{r}
library(dplyr)
library(ggplot2)
library(reshape2)
library(ggrepel)
library(scales)
library(ggpubr)
source('~/HexamerBias/artificial_coverage/compare_peaks.r')

prevalent_nucleotide <- function(seq){
  nuc.count <- table(strsplit(seq, ''))
  prev.nuc <- names(which.max(nuc.count))
  return(prev.nuc)
}
```

# Primer usage in same samples
Load data for D2
```{r, echo=F}
setwd("~/mnt/edann/primer_conc_VAN2493/")
D2G <- loadPtMatrix("~/mnt/edann/primer_conc_VAN2493/D2G_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
D2T <- loadPtMatrix("~/mnt/edann/primer_conc_VAN2493/D2T_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
D2R <- loadPtMatrix("~/mnt/edann/primer_conc_VAN2493/D2R_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
```

```{r, echo=F}
d2g.primer.usage <- make.hex.usage.df(D2G, type="primer", scale=TRUE)
d2t.primer.usage <- make.hex.usage.df(D2T, type="primer", scale=TRUE)
d2r.primer.usage <- make.hex.usage.df(D2R, type="primer", scale=TRUE)
```

```{r, echo=F}
format.cor.plot <- function(plot){
  plot <- plot + 
    theme_classic() +
    geom_point(aes(color=top.nucleotide), alpha=0.4) +
    # geom_text_repel() +
    geom_abline(intercept = 0, slope=1, color='red') +
    theme(axis.title = element_text(size = 28), 
          axis.text = element_text(size=20),
          legend.text = element_text(size=20),
          legend.title = element_blank()) 
  return(plot)
}

d2.primer.df <- inner_join(d2g.primer.usage, d2t.primer.usage, by='primer') %>%
  inner_join(.,d2r.primer.usage, by='primer') %>%
  rename(primer.usage.T= D2T,
         primer.usage.G= D2G,
         primer.usage.R= D2R,) %>%
  mutate(top.nucleotide=sapply(as.character(primer), prevalent_nucleotide))

p1 <- d2.primer.df %>%  
  ggplot(., aes(primer.usage.G, primer.usage.T))  
p2 <- d2.primer.df %>%  
  ggplot(., aes(primer.usage.R, primer.usage.T))  
p3 <- d2.primer.df %>%  
  ggplot(., aes(primer.usage.R, primer.usage.G)) 


format.cor.plot(p1) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/primer_cor_GT.pdf", width = 12, height = 10)
format.cor.plot(p2) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/primer_cor_RT.pdf", width = 12, height = 10)
format.cor.plot(p3) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/primer_cor_RG.pdf", width = 12, height = 10)
```

# Template usage in same samples
```{r, echo=T}
d2g.template.usage <- make.hex.usage.df(D2G, type="template", scale=TRUE)
d2t.template.usage <- make.hex.usage.df(D2T, type="template", scale=TRUE)
d2r.template.usage <- make.hex.usage.df(D2R, type="template", scale=TRUE)

d2.template.df <- inner_join(d2g.template.usage, d2t.template.usage, by='template') %>%
  inner_join(.,d2r.template.usage, by='template') %>%
  rename(template.usage.T= D2T,
         template.usage.G= D2G,
         template.usage.R= D2R,) %>%
  mutate(top.nucleotide=sapply(as.character(template), prevalent_nucleotide))
```

```{r, echo=F}
p1 <- d2.template.df %>%  
  ggplot(., aes(template.usage.G, template.usage.T))  
p2 <- d2.template.df %>%  
  ggplot(., aes(template.usage.R, template.usage.T))  
p3 <- d2.template.df %>%  
  ggplot(., aes(template.usage.R, template.usage.G)) 


format.cor.plot(p1) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/template_cor_GT.pdf", width = 12, height = 10)
format.cor.plot(p2) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/template_cor_RT.pdf", width = 12, height = 10)
format.cor.plot(p3) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/template_cor_RG.pdf", width = 12, height = 10)
```

# Cumsum of hexamer and template usage
```{r}
d <- data.frame(G.cumsum = cumsum(sort(d2.primer.df$D2G, decreasing = TRUE)), 
                T.cumsum = cumsum(sort(d2.primer.df$D2T, decreasing = TRUE)),
                R.cumsum = cumsum(sort(d2.primer.df$D2R, decreasing = TRUE)))
d1 <- d %>% mutate(ix=as.numeric(row.names(d))) %>%
  melt(variable.name='sample', id.vars=c('ix')) 
ggplot(d1, aes(ix,value, color=sample)) + geom_point() +   
  ggtitle('Cum distribution primer sequences - D2') +
  theme(axis.title = element_text(size = 22), 
        axis.text = element_text(size=14),
        title = element_text(size=22)) 
```

```{r}
d <- data.frame(G.cumsum = cumsum(sort(d2.template.df$D2G, decreasing = TRUE)), 
                T.cumsum = cumsum(sort(d2.template.df$D2T, decreasing = TRUE)),
                R.cumsum = cumsum(sort(d2.template.df$D2R, decreasing = TRUE)))
d1 <- d %>% mutate(ix=as.numeric(row.names(d))) %>%
  melt(variable.name='sample', id.vars=c('ix')) 
ggplot(d1, aes(ix,value, color=sample)) + geom_point() +   
  ggtitle('Cum distribution template sequences - D2') +
  theme(axis.title = element_text(size = 22), 
        axis.text = element_text(size=14),
        title = element_text(size=22)) 
```
# Distribution of coverage per template sequence

*Absolute template coverage*
```{r, echo=T}
d2g.template.abs <- make.hex.usage.df(D2G, type="template", scale=F)
d2t.template.abs <- make.hex.usage.df(D2T, type="template", scale=F)
d2r.template.abs <- make.hex.usage.df(D2R, type="template", scale=F)

d2.template.abs.df <- inner_join(d2g.template.abs, d2t.template.abs, by='template') %>%
  inner_join(.,d2r.template.abs, by='template')

d2.template.abs.df %>%
  melt(variable.name='sample', value.name='template.usage') %>%
  ggplot(., aes(sample,template.usage)) + 
  # facet_grid(sample~.) +
  geom_boxplot()

d2.template.abs.df %>%
  melt(variable.name='sample', value.name='template.usage') %>%
  ggplot(., aes(sample,template.usage)) + 
  # facet_grid(sample~.) +
  geom_boxplot(outlier.shape = NA) +
  ylim(0,1000) 
  
```

*Absolute primer coverage*
```{r, echo=T}
d2g.primer.abs <- make.hex.usage.df(D2G, type="primer", scale=F)
d2t.primer.abs <- make.hex.usage.df(D2T, type="primer", scale=F)
d2r.primer.abs <- make.hex.usage.df(D2R, type="primer", scale=F)

d2.primer.abs.df <- inner_join(d2g.primer.abs, d2t.primer.abs, by='primer') %>%
  inner_join(.,d2r.primer.abs, by='primer')

d2.primer.abs.df %>%
  melt(variable.name='sample', value.name='primer.usage') %>%
  ggplot(., aes(sample,primer.usage)) + 
  # facet_grid(sample~.) +
  geom_boxplot()

d2.primer.abs.df %>%
  melt(variable.name='sample', value.name='primer.usage') %>%
  ggplot(., aes(sample,primer.usage)) + 
  # facet_grid(sample~.) +
  geom_boxplot(outlier.shape = NA) +
  ylim(0,1000) 
  
```
## Same but with different samples
```{r}
setwd("~/mnt/edann/primer_conc_VAN2493/")
P1G <- loadPtMatrix("P1G_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
P1T <- loadPtMatrix("P1T_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
P1R <- loadPtMatrix("P1R_tr2_R1_bismark_bt2.ptCounts.qualFilt.csv",compression = 'none')
```

```{r, echo=T}
p1g.primer.usage <- make.hex.usage.df(P1G, type="primer", scale=TRUE)
p1t.primer.usage <- make.hex.usage.df(P1T, type="primer", scale=TRUE)
p1r.primer.usage <- make.hex.usage.df(P1R, type="primer", scale=TRUE)
```

```{r, echo=F}
p1.primer.df <- inner_join(p1g.primer.usage, p1t.primer.usage, by='primer') %>%
  inner_join(.,p1r.primer.usage, by='primer')

p1.primer.df %>%  
  mutate(label = ifelse(P1G > 0.008 | P1T > 0.005, as.character(primer),'')) %>%
  ggplot(., aes(P1G, P1T, label=label)) + 
  theme_classic() +
  geom_point() +
  geom_text_repel() +
  theme(axis.title = element_text(size = 22), 
        axis.text = element_text(size=14))

p1.primer.df %>%  
  mutate(label = ifelse(P1G > 0.008 | P1T > 0.005, as.character(primer),'')) %>%
  ggplot(., aes(P1G, P1R, label=label)) + 
  theme_classic() +
  geom_point() +
  geom_text_repel() +
  theme(axis.title = element_text(size = 22), 
        axis.text = element_text(size=14))

p1.primer.df %>%  
  mutate(label = ifelse(P1R > 0.008 | P1T > 0.005, as.character(primer),'')) %>%
  ggplot(., aes(P1T, P1R, label=label)) + 
  theme_classic() +
  geom_point() +
  geom_text_repel() +
  theme(axis.title = element_text(size = 22), 
        axis.text = element_text(size=14))
```

# Coverage bias 
```{r}
source("~/HexamerBias/quantify_bias/biasProfiles.r")

d2t.mat <- load.matrix("D2T_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")
d2r.mat <- load.matrix("D2R_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")
d2g.mat <- load.matrix("D2G_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")


df.cov <- make.df.of.profiles(list(moreT=d2t.mat,
                                   moreG=d2g.mat,
                                   random=d2r.mat
                                   ))

p <- plot.genes.profile.df(df.cov, big.labels =F)
p 
```


```{r}
setwd("~/mnt/edann/primer_conc_VAN2493/")
d3t.mat <- load.matrix("D3T_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")
d3r.mat <- load.matrix("D3R_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")
d3g.mat <- load.matrix("D3G_tr2_R1_bismark_bt2.deduplicated.srt.mat.gz")


df.cov <- make.df.of.profiles(list(moreT=d3t.mat,
                                   moreG=d3g.mat,
                                   random=d3r.mat
                                   ))

p <- plot.genes.profile.df(df.cov, big.labels =F)
p
```

## PREDICTED COVERAGE ADDING BATCH INFORMATION
Load predicted coverage (made with ```compute_predicted_coverage```)
```{r}
G_predcov <- read.csv("~/mnt/edann/primer_conc_VAN2493/DR_G_ppm.csv.gz.predcoverage.csv")
R_predcov <- read.csv("~/mnt/edann/primer_conc_VAN2493/DR_R_ppm.predcoverage.csv")
T_predcov <- read.csv("~/mnt/edann/primer_conc_VAN2493/DR_T_ppm.csv.gz.predcoverage.csv")
```
Correlate
```{r}
plot_coverage_prediction <- function(predcov, template.usage, obs.label='', exp.label=''){
  pred.df <- predcov %>%
    mutate(expected=exp/sum(exp)) %>%
    inner_join(., template.usage, by='template') %>%
    select(template, expected, observed) 
  pcc <- cor(pred.df$expected, pred.df$observed, use='pairwise.complete.obs')
  pl <- pred.df %>%
    mutate(top.nucleotide=sapply(as.character(template), prevalent_nucleotide)) %>% 
    # mutate(label=ifelse(expected>0.002 | observed>0.002, as.character(template), '')) %>%
    ggplot(., aes(expected, observed, color=top.nucleotide)) +
    geom_point(alpha=0.5, size=2.5) +
    xlab(paste('expected', exp.label)) + ylab(paste('observed', obs.label)) +
    # geom_text_repel() +
    ggtitle(paste('PCC =', round(pcc,3))) +
    geom_abline(intercept = 0, slope=1, color='red') +
    theme_classic() 
    # theme(axis.title = element_text(size = 32),
    #   axis.text = element_text(size=28),
    #   legend.text = element_text(size=35),
    #   legend.title = element_blank(),
    #   title=element_text(size=32)) 
  return(pl)
}

plot_coverage_prediction(G_predcov, rename(d2g.template.usage,observed=D2G), exp.label = 'G', obs.label='G')
plot_coverage_prediction(R_predcov, rename(d2r.template.usage,observed=D2R), exp.label = 'R', obs.label='R')
plot_coverage_prediction(T_predcov, rename(d2t.template.usage,observed=D2T), exp.label = 'T', obs.label='T')
```

Residual plots
```{r}
plot.residuals <- function(predcov, template.usage, obs.label='', exp.label=''){
  pred.df <-predcov %>%
    mutate(expected=exp/sum(exp)) %>%
    inner_join(., template.usage, by='template') %>%
    select(template, expected, observed) 
  pl <- pred.df %>%
    mutate(top.nucleotide=sapply(as.character(template), prevalent_nucleotide)) %>% 
    # mutate(observed=log(observed), expected=log(expected)) %>%
    # mutate(label=ifelse(expected>0.002 | observed>0.002, as.character(template), '')) %>%
    ggplot(., aes(observed, observed-expected, color=top.nucleotide)) +
    geom_point(alpha=0.5, size=2.5) +
    xlab(paste('obs', exp.label)) + ylab(paste('obs-exp', obs.label)) +
    # geom_text_repel() +
    geom_hline(yintercept = 0, color='red') +
    theme_classic() 
  return(pl)    
  }

plot.residuals(G_predcov, rename(d2g.template.usage,observed=D2G), exp.label = 'G', obs.label='G')
plot.residuals(R_predcov, rename(d2r.template.usage,observed=D2R), exp.label = 'R', obs.label='R')
plot.residuals(T_predcov, rename(d2t.template.usage,observed=D2T), exp.label = 'T', obs.label='T')

```
Random mix
```{r}
plot_coverage_prediction(G_predcov, rename(d2t.template.usage,observed=D2T), exp.label = 'G', obs.label='T')
plot_coverage_prediction(T_predcov, rename(d2g.template.usage,observed=D2G), exp.label = 'T', obs.label='G')

plot_coverage_prediction(G_predcov, rename(d2r.template.usage,observed=D2R), exp.label = 'G', obs.label='R')
plot_coverage_prediction(T_predcov, rename(d2r.template.usage,observed=D2R), exp.label = 'T', obs.label='R')

plot_coverage_prediction(R_predcov, rename(d2t.template.usage,observed=D2T), exp.label = 'R', obs.label='T')
plot_coverage_prediction(R_predcov, rename(d2g.template.usage,observed=D2G), exp.label = 'R', obs.label='G')
```
```{r}
inner_join(R_predcov, G_predcov, by='template', suffix=c(".R", ".G")) %>%
  mutate(label=sapply(as.character(template), prevalent_nucleotide)) %>%
  # mutate(label=ifelse(grepl(template, pattern = 'GGG'),'yes', 'no')) %>%
  ggplot(., aes(exp.R, exp.G, color=label)) +
  geom_point(alpha=0.4) +
  theme(axis.text = element_text(size=20)) +
  # geom_text_repel() +
  geom_abline(intercept = 0, slope=1)

inner_join(R_predcov, T_predcov, by='template', suffix=c(".R", ".T")) %>%
  mutate(label=sapply(as.character(template), prevalent_nucleotide)) %>% 
  ggplot(., aes(exp.R, exp.T, color=label)) +
  geom_point(alpha=0.4) +
  theme(axis.text = element_text(size=20)) +
  # geom_text_repel() +
  geom_abline(intercept = 0, slope=1)

```

# Correlation of deltaG
```{r}
G.deltaG <- loadDgMatrix('~/mnt/edann/primer_conc_VAN2493/D2G_tr2_R1_bismark_bt2_ptDg_qual.csv', compression = 'none')
T.deltaG <- loadDgMatrix('~/mnt/edann/primer_conc_VAN2493/D2T_tr2_R1_bismark_bt2_ptDg_qual.csv', compression = 'none')
R.deltaG <- loadDgMatrix('~/mnt/edann/primer_conc_VAN2493/D2R_tr2_R1_bismark_bt2_ptDg_qual.csv', compression = 'none')

G.pair <- make_pair_df(G.deltaG)
T.pair <- make_pair_df(T.deltaG)
R.pair <- make_pair_df(R.deltaG)

G.pair.diag <- G.pair[substr(G.pair$ptPair,1,6)==substr(G.pair$ptPair,8,13),]
R.pair.diag <- R.pair[substr(R.pair$ptPair,1,6)==substr(R.pair$ptPair,8,13),]
T.pair.diag <- T.pair[substr(T.pair$ptPair,1,6)==substr(T.pair$ptPair,8,13),]
```

```{r}
common.diag <- inner_join(R.pair.diag, T.pair.diag, by='ptPair', suffix=c('.R', '.T')) %>% 
  filter(!is.infinite(dG.R) & !is.infinite(dG.T))

common.diag %>%
  mutate(nuc=sapply(ptPair, prevalent_nucleotide)) %>%
  ggplot(., aes(dG.R, dG.T)) +
    geom_point(alpha=0.3, aes(color=nuc))

common.diag <- inner_join(G.pair.diag, T.pair.diag, by='ptPair', suffix=c('.G', '.T')) %>% 
  filter(!is.infinite(dG.G) & !is.infinite(dG.T))

common.diag %>%
  mutate(nuc=sapply(ptPair, prevalent_nucleotide)) %>%
  ggplot(., aes(dG.G, dG.T)) +
    geom_point(alpha=0.3, aes(color=nuc))


```


```{r}
make_ppm_of_usage <- function(primer.usage.df){
  df <- primer.usage.df
  seqList <- unlist(lapply(1:nrow(df), function(i) rep(as.character(df[i,1]), df[i,2])))
  mat <- apply(do.call(rbind,strsplit(seqList, split='')),2,table)
  prob.mat <- apply(mat,2,function(x) x/sum(x))
  pwm <- makePWM(prob.mat)
  return(pwm)
}
pwm.primer.g <- make_ppm_of_usage(d2g.primer.abs)
pwm.primer.t <- make_ppm_of_usage(d2t.primer.abs)
pwm.primer.r <- make_ppm_of_usage(d2r.primer.abs)
as.data.frame(pwm.primer.g@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('')
as.data.frame(pwm.primer.t@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('')
as.data.frame(pwm.primer.r@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('')
```

```{r}
pwm.template.g <- make_ppm_of_usage(d2g.template.abs)
pwm.template.t <- make_ppm_of_usage(d2t.template.abs)
pwm.template.r <- make_ppm_of_usage(d2r.template.abs)
as.data.frame(pwm.template.g@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('') + ggsave('~/AvOwork/output/primer_batch_VAN2493/template_usage_ppm_G.pdf')
as.data.frame(pwm.template.t@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('') + ggsave('~/AvOwork/output/primer_batch_VAN2493/template_usage_ppm_T.pdf')
as.data.frame(pwm.template.r@pwm) %>% mutate(nuc=rownames(.)) %>% melt(id.vars='nuc', value.name='prob', variable.name='pos') %>% plot.nuc.matrix(.) + ggtitle('') + ggsave('~/AvOwork/output/primer_batch_VAN2493/template_usage_ppm_R.pdf')

```

# How many CpG sites do I cover with different primer concentrations?
```{r}
cpg.tab <- read.table("~/mnt/edann/primer_conc_VAN2493/cpgs_covered.txt", col.names = c('smp', 'n_cpgs'))
cpg.tab <- cpg.tab %>%
  mutate(batch=substr(smp, 3,4),
         smp=substr(smp,1,2)) 
cpg.tab %>%
  ggplot(., aes(smp, n_cpgs, group=batch, fill=batch)) +
  geom_bar(stat='identity', position='dodge')

```

Number of CpGs vs number of reads
```{r}
nreads.tab <- read.table("~/mnt/edann/primer_conc_VAN2493/nreads_bam.txt", col.names = c('smp', 'nreads'))
tab <- nreads.tab %>%
  mutate(batch=substr(smp, 3,4),
         smp=substr(smp,1,2)) %>%
  inner_join(., cpg.tab, by=c('smp', 'batch'))
tab %>%
  arrange(nreads)
```

```{r}
myColors <- brewer.pal(3,"Accent")
names(myColors) <- c("G", "T", "R")
col.scale.primer.batch <- scale_colour_manual(name = "",values = myColors)

p3 <- tab %>%
  ggplot(., aes(nreads, n_cpgs, group=batch, color=batch)) +
  # geom_line() +
  geom_point(size=1.5) +
  col.scale.primer.batch +
  geom_line(stat='smooth', method='lm', se=F, size=2, alpha=0.5) +
  scale_x_continuous(labels=scientific) + scale_y_continuous(labels=scientific) +
  theme_minimal() +
  stat_smooth_func(geom="text",method="lm",se = F,xpos = 1.5e+06, parse=T)
  # stat_summary() +
  xlab("# reads") + ylab("# covered CpGs") +
  theme(legend.title = element_blank(),
        legend.text = element_text(size=20),
        axis.text = element_text(size=16),
        axis.title = element_text(size=30),
        title = element_text(size=30)) 

p3
  group_by(batch) %>%
  stat_summ

p3 + ggsave("~/AvOwork/output/primer_batch_VAN2493/nreads_VS_ncpgs.pdf")
```

## Predicted coverage profiles 
```{r}
source("~/HexamerBias/quantify_bias/biasProfiles.r")

dt.mat <- load.matrix("~/mnt/edann/primer_conc_VAN2493/mm10.highcov.rand.42DR_T_ppm.artCov.mat.gz")
dr.mat <- load.matrix("~/mnt/edann/primer_conc_VAN2493/mm10.highcov.rand.42DR_R_ppm.predcoverage.artCov.mat.gz")
dg.mat <- load.matrix("~/mnt/edann/primer_conc_VAN2493/mm10.highcov.rand.42DR_G_ppm.artCov.mat.gz")

df.cov <- make.df.of.profiles(list(moreT=dt.mat,
                                   moreG=dg.mat,
                                   random=as.vector(smooth(dr.mat))
                                   ))

p <- plot.genes.profile.df(df.cov, big.labels =F)
p
```
```{r}
all.profiles.bs <- list(moreT=dt.mat,
                     moreG=dg.mat,
                     random=dr.mat)
                     
df <- as.data.frame(all.profiles.bs)[300,] %>%
  gather(key = 'sample', value = 'cov') %>%
  mutate(batch=sample) 

myColors <- brewer.pal(3,"Accent")
names(myColors) <- c("moreG", "moreT", "random")
fill.scale.primer.batch <- scale_fill_manual(name = "",values = myColors)
color.scale.primer.batch <- scale_color_manual(name = "",values = myColors)

df %>%
  mutate(batch=factor(batch,levels = c('moreG', 'random', 'moreT'))) %>%
  # factor(x$name, levels = x$name[order(x$val)])
  ggplot(., aes(batch, cov)) +
  geom_bar(aes(fill=batch), stat='identity') +
  theme_classic() +
  fill.scale.primer.batch +
  ylab("Coverage (z-score)") + xlab("") +
  ggtitle("TSS predicted coverage") +
      theme(axis.title = element_text(size = 25), 
          axis.text = element_text(size=18), 
          plot.title = element_text(hjust = 0.5,size=33),
          legend.text = element_text(size=22), 
          legend.key.size = unit(1,"cm"),
          panel.spacing = unit(2, "lines"),
          strip.text.x = element_text(size = 20)) +
  theme(axis.text.x = element_blank()) +
  ggsave("~/AvOwork/output/primer_batch_VAN2493/predicted_coverage_TSS.pdf")
```

## Methylation bias with different batches
I have computed the avg methylation in a set of random regions for all the samples
```{bash}
bedtools random -n 10000000 -l 100 -g ../genomes/mm10/mm10.genome > random_mm10.bed
for file in $(ls *.cov.gz); 
  do 
  zcat $file | bedtools intersect -a random_mm10.bed -b stdin -wo | bedtools groupby -g 1,2,3 -c 10 -o mean | awk -v file="$file" '{print $0"\t"substr(file,1,3)}'; 
  done > meth_random_bigger.bed
```

```{r}
met <- read.table("~/mnt/edann/primer_conc_VAN2493/meth_random_bigger.bed", col.names = c('chr', 'start', 'end', 'met.frac', 'sample'))
batch.met <- met %>%
  filter(grepl(sample, pattern = 'D.+')) %>%
  mutate(region=paste(chr,start,end, sep='.'),
         batch=substr(sample,3,4)) %>%
  dplyr::select(region,batch,met.frac) %>%
  group_by(batch, region) %>%
  summarize(met.frac=mean(met.frac, na.rm=T)) %>%
  dcast(region ~ batch) 
```

```{r}
batch.met %>%
  ggplot(., aes(T,G)) + geom_point(alpha=0.3)

batch.met %>%
  # filter(!is.na(G) & !is.na(T) & !is.na(R)) %>%
  melt %>%
  group_by(variable) %>%
  arrange(value) %>%
  filter(!is.na(value)) %>%
  mutate(rank=rank(value)) %>%
  ungroup %>%
  dplyr::select(-region) %>%
  # arrange(T) %>%
  # mutate(rank.T=1:n()) %>%
  # # dplyr::select(T, rank.T)
  # arrange(R) %>%
  # mutate(rank.R=1:n()) %>%
  # arrange(G) %>%
  # mutate(rank.G=1:n()) %>%
  # # sample_n(10) %>%
  # dplyr::select(-region) 
  # melt(id.vars=c('G', 'T', 'R')) 
  ggplot(., aes(rank,value, group=variable, color=variable)) + geom_line(alpha=0.8)
cor(dplyr::select(batch.met, G,R,T), use='pairwise.complete.obs')
```

```{r}
batch.met %>%
  melt %>%
  ggplot(., aes(variable, value, color=variable)) + geom_violin(alpha=0.3)
```


```{r}
smp.met <- met %>%
  mutate(region=paste(chr,start,end, sep='.'),
         batch=substr(sample,3,4)) %>%
  # filter(grepl(sample, pattern = 'D.+')) %>%
  group_by(sample, region) %>%
  summarize(met.frac=mean(met.frac, na.rm=T)) %>%
  dcast(region ~ sample) 

cor.smp.met <- cor(dplyr::select(smp.met, -region), use='pairwise.complete.obs')
cor.smp.met[cor.smp.met==1] <- NA
hm <- pheatmap(cor.smp.met,
               annotation_col = anno.df, annotation_row = anno.df,
               cellheight = 10, cellwidth = 10,
               filename = "~/AvOwork/formatted_figs/suppl_figs_final/met_batches_all.pdf",
               width = 15, height=10)


anno.df <- data.frame(smp=colnames(smp.met[-1]) ) %>%
  mutate(sample=substr(smp,1,1),
         batch= substr(smp,3,3),
         rep=substr(smp,1,2)
         ) %>%
  column_to_rownames(var = 'smp')

## Top variance regions
var.regions <- smp.met %>% 
  na.omit() %>% 
  melt(id.vars='region', variable.name='sample', value.name='met') %>% 
  group_by(region) %>% 
  summarise(var=var(met)) %>% 
  top_n(1000, var)

high.var.regs <- smp.met %>%
  filter(region %in% var.regions$region) 

pheatmap(high.var.regs[-1],
        show_rownames = F,
        annotation_col = anno.df,
        filename = "~/AvOwork/formatted_figs/suppl_figs_final/met_batches_all.pdf") 

corr <- cor(dplyr::select(smp.met, -region), use='pairwise.complete.obs')
corr[corr==1] <- NA
pheatmap::pheatmap(corr)
```

------
Load profiles and save as RData
```{r}
setwd("~/mnt/edann/primer_conc_VAN2493/")
dt.coverage <- load.expVSpred.coverage("mm10.highcov.rand.42DR_T_ppm.artCov.bw", "mm10.highcov.rand.42DR_G_ppm.artCov.bw")
l.dt.coverage <- split(dt.coverage, dt.coverage$id)
norm.l.dt.coverage <- sapply(l.dt.coverage, normalize.coverage)
```

```{r}

```
No sense. Too little coverage.
