---
title: "Chi-square approach validation"
output: html_notebook
---
```{r, echo=F}
source("~/HexamerBias/deltaGprediction/binding_model_functions.r")
```

Validation of chi-square approach to estimate the scaling factor $\epsilon$ and the equilibrium constant for binding of matching pairs $K_{eq}$, using different genomes.

# Loading
```{r}
cele.df <- make.match.df("~/mnt/edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-CeleTotal-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv", '~/mnt/edann/hexamers/genomes_kmers/WBcel235.kmerAbundance.csv')
zf.df <- make.match.df("~/mnt/edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-zfishTotal-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv", '~/mnt/edann/hexamers/genomes_kmers/danRer10.kmerAbundance.csv')
human.df <- make.match.df("~/mnt/edann/VAN2423_onePreamp/cov_prediction/CG-pbat-gDNA-humanAPKS-noBS-1preAmp-handMix_lmerged_R1.ptCounts.qualFilt.csv", '~/mnt/edann/hexamers/genomes_kmers/hg38.kmerAbundance.csv')

```

(Saving RData cos it's about time)
```{r}
save(cele.df, file='~/AvOwork/cele_pt.RData')
save(zf.df, file='~/AvOwork/zf_pt.RData')
save(human.df, file='~/AvOwork/human_pt.RData')
```


Build primer pool for random hexamers
```{r}
primer.pool <- simulate.primer.pool(pool.size = 200000)
hist(table(primer.pool))
```

Load groups of sequences with same binding energy
```{r}
t.groups <- read.table("~/AvOwork/MatchingTemplate_groups.tsv", col.names = c('template', 'group'))
```

# Estimate epsilon for every experiment
For every genome sample 20 groups with at least 20 sequences, check distributions within samples and the total distribution

```{r}
epsilon.iterative <- function(pt.df, estimation.it=10, tot.it=20, sample.size=20, imposeP=F){
  i<-1
  eps.df <- data_frame(n=1:estimation.it, estimate.epsilon(pt.df, primer.pool, sample(seq(1,40),sample.size), iterations = estimation.it, imposeP=imposeP))
  colnames(eps.df)[ncol(eps.df)] <- paste0('it.',i)
  while(i<tot.it){
    i<-i+1
    eps.df <- full_join(eps.df, data_frame(n=1:estimation.it, estimate.epsilon(pt.df, primer.pool, sample(seq(1,40),sample.size), iterations = estimation.it, imposeP=imposeP)), by='n')
    colnames(eps.df)[ncol(eps.df)] <- paste0('it.',i)
  }
  return(eps.df)
  }

cele.eps <-epsilon.iterative(cele.df, estimation.it = 20)
zf.eps <-epsilon.iterative(zf.df, estimation.it = 20)
human.eps <-epsilon.iterative(human.df, estimation.it = 20)

cele.eps %>%
  melt(id.vars='n', variable.name='iteration', value.name = 'eps') %>%
  ggplot(., aes(iteration, eps)) + geom_boxplot()

zf.eps %>%
  melt(id.vars='n', variable.name='iteration', value.name = 'eps') %>%
  ggplot(., aes(iteration, eps)) + geom_boxplot()

human.eps %>%
  melt(id.vars='n', variable.name='iteration', value.name = 'eps') %>%
  ggplot(., aes(iteration, eps)) + geom_boxplot()

```

Compare epsilon estimation using probability distribution or mean probability
```{r}
zf.eps <-epsilon.iterative(zf.df, estimation.it = 20, sample.size = 10)
zf.eps.prob <-epsilon.iterative(zf.df, estimation.it = 50, sample.size = 10, imposeP = T)
cele.eps.prob <-epsilon.iterative(cele.df, estimation.it = 50, sample.size = 10, imposeP = T)
median(as.matrix(zf.eps))
median(as.matrix(zf.eps.prob))
sd(as.matrix(zf.eps))
sd(as.matrix(zf.eps.prob))

median(as.matrix(cele.eps))
median(as.matrix(cele.eps.prob))
sd(as.matrix(cele.eps))
sd(as.matrix(cele.eps.prob))

# cbind(c(as.matrix(cele.eps.prob)), c(as.matrix(zf.eps.prob)), c(as.matrix(cele.eps)), c(as.matrix(zf.eps))) %>%

mutate(melt(cele.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele.dist') %>%
  bind_rows(., mutate(melt(cele.eps.prob, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele.prob')) %>%
  # bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>%
  ggplot(., aes(eps, group=species, fill=species)) + 
  facet_grid(species~.) +
  geom_histogram()  
  # group_by(species) %>%
  # summarise(mean(eps, na.rm=T), sd(eps, na.rm = T))
  #   ggplot(., aes(species,eps)) + geom_boxplot()

mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf.dist') %>%
  bind_rows(., mutate(melt(zf.eps.prob, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf.prob')) %>%
  # bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>%
  ggplot(., aes(eps, group=species, fill=species)) + 
  facet_grid(species~.) +
  geom_histogram()  
```

Effect of sample size on estimated $\varepsilon$
```{r}
cele.eps.3 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 3)
cele.eps.4 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 4)
cele.eps.5 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 5)
cele.eps.10 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 10)
cele.eps.15 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 15)
cele.eps.20 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 20)
cele.eps.25 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 25)
cele.eps.30 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 30)

mutate(melt(cele.eps.5, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=5) %>%
  # bind_rows(., mutate(melt(cele.eps.3, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=3)) %>%
  # bind_rows(., mutate(melt(cele.eps.4, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=4)) %>%
  bind_rows(., mutate(melt(cele.eps.10, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=10)) %>%
  bind_rows(., mutate(melt(cele.eps.15, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=15)) %>%
  bind_rows(., mutate(melt(cele.eps.20, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=20)) %>%
  bind_rows(., mutate(melt(cele.eps.25, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=25)) %>%
  bind_rows(., mutate(melt(cele.eps.30, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=30)) %>%
# bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>%
  ggplot(., aes(sample.size, eps, group=sample.size, fill=sample.size)) + 
  # facet_grid(species~.) +
  geom_violin()  
```
```{r}
cele.eps.3 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 3,imposeP = T)
cele.eps.4 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 4,imposeP = T)
cele.eps.5 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 5,imposeP = T)
cele.eps.10 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 10,imposeP = T)
cele.eps.15 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 15,imposeP = T)
cele.eps.20 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 20,imposeP = T)
cele.eps.25 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 25,imposeP = T)
cele.eps.30 <-epsilon.iterative(select.diag.pairs(cele.all.df), estimation.it = 20, sample.size = 30,imposeP = T)

mutate(melt(cele.eps.5, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=5) %>%
  # bind_rows(., mutate(melt(cele.eps.3, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=3)) %>%
  # bind_rows(., mutate(melt(cele.eps.4, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=4)) %>%
  bind_rows(., mutate(melt(cele.eps.10, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=10)) %>%
  bind_rows(., mutate(melt(cele.eps.15, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=15)) %>%
  bind_rows(., mutate(melt(cele.eps.20, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=20)) %>%
  bind_rows(., mutate(melt(cele.eps.25, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=25)) %>%
  bind_rows(., mutate(melt(cele.eps.30, id.vars='n', variable.name='iteration', value.name = 'eps'), sample.size=30)) %>%
# bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>%
  ggplot(., aes(sample.size, eps, group=sample.size, fill=sample.size)) + 
  # facet_grid(species~.) +
  geom_violin()  
```

Total distribution
```{r}
mutate(melt(cele.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='cele') %>%
  bind_rows(., mutate(melt(human.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='human')) %>%
  bind_rows(., mutate(melt(zf.eps, id.vars='n', variable.name='iteration', value.name = 'eps'), species='zf')) %>%
  ggplot(., aes(eps, group=species, fill=species)) + geom_histogram()  
  group_by(species) %>%
  summarise(mean(eps, na.rm=T), sd(eps, na.rm = T))
    ggplot(., aes(species,eps)) + geom_boxplot()

    

```

One reassuring thing is that I get a much higher epsilon for the c.elegans
```{r}
# keq.iterative <- function(pt.df, estimation.it=10, sample.size=20){
#   i<-1
#   keq.df <- data_frame(n=1:estimation.it, estimate.epsilon(pt.df, primer.pool, sample(seq(1,40),sample.size), iterations = estimation.it))
#   colnames(eps.df)[ncol(eps.df)] <- paste0('it.',i)
#   while(i<tot.it){
#     i<-i+1
#     eps.df <- full_join(eps.df, data_frame(n=1:estimation.it, estimate.epsilon(pt.df, primer.pool, sample(seq(1,40),sample.size), iterations = estimation.it)), by='n')
#     colnames(eps.df)[ncol(eps.df)] <- paste0('it.',i)
#   }
#   return(eps.df)
#   }

keq.cele <-estimate.keq(cele.df, primer.pool, groupsOI = seq(1,20), iterations = 100)
keq.human <-estimate.keq(human.df, primer.pool, groupsOI = c(1,2,4,5,6,7,8, seq(9,20)), iterations = 100)
keq.zf <-estimate.keq(zf.df, primer.pool, groupsOI = seq(1,20), iterations = 100)
keq.cele %>%
  melt %>%
  ggplot(., aes(variable, value, fill=variable)) + geom_boxplot()

# as.data.frame(keq.human) %>%
#   melt %>%
#   ggplot(., aes(variable, value, fill=variable)) + geom_boxplot()
# 
as.data.frame(keq.zf) %>%
  melt %>%
  ggplot(., aes(variable, value, fill=variable)) + geom_boxplot()

```
```{r}
compare.zf.cele <- mutate(melt(keq.cele, variable.name='group',value.name = 'keq'), species='cele') %>%
  bind_rows(., mutate(melt(keq.human,variable.name='group',value.name = 'keq' ) , species='human')) %>%
  bind_rows(., mutate(melt(keq.zf, variable.name='group',value.name = 'keq'), species='zf')) 

compare.zf.cele %>%
  ggplot(., aes(group, keq, fill=species)) + geom_boxplot()

compare.zf.cele %>%
  group_by(group, species) %>%
  summarise(mean.keq=mean(keq)) %>%
  dcast(group ~ species) %>%
  ggplot(., aes(zf, human, label=group)) + geom_text() +
  geom_abline(intercept = 0, slope=1, color='red') +
  theme_classic() +
  geom_errorbar(aes(ymin=human+sd(human, na.rm=T), ymax=human-sd(human, na.rm=T)), alpha=0.4) +
  geom_errorbarh(aes(xmin=zf+sd(zf, na.rm=T), xmax=zf-sd(zf, na.rm=T)), alpha=0.4)

compare.zf.cele %>%
  group_by(group, species) %>%
  summarise(mean.keq=mean(keq)) %>%
  dcast(group ~ species) %>%
  ggplot(., aes(zf, cele, label=group)) + geom_text() +
  geom_abline(intercept = 0, slope=1, color='red') +
  theme_classic()+
  geom_errorbar(aes(ymin=cele+sd(cele, na.rm=T), ymax=cele-sd(cele, na.rm=T)), alpha=0.4) +
  geom_errorbarh(aes(xmin=zf+sd(zf, na.rm=T), xmax=zf-sd(zf, na.rm=T)), alpha=0.4) 
  
compare.zf.cele %>%
  group_by(group, species) %>%
  summarise(mean.keq=mean(keq)) %>%
  dcast(group ~ species) %>%
  ggplot(., aes(cele, human, label=group)) + geom_text() +
  theme_classic()+
  geom_abline(intercept = 0, slope=1, color='red') +
  geom_errorbar(aes(ymin=human+sd(human, na.rm=T), ymax=human-sd(human, na.rm=T)), alpha=0.4) +
  geom_errorbarh(aes(xmin=cele+sd(cele, na.rm=T), xmax=cele-sd(cele, na.rm=T)), alpha=0.4)

```
# Estimate deltaG
Using info from primer concentration assigned from simulated distribution
```{r}
deltaG.zf <-estimate.deltaG(zf.df, primer.pool, groupsOI = seq(1,20), iterations = 100)
deltaG.cele <-estimate.deltaG(cele.df, primer.pool, groupsOI = seq(1,20), iterations = 100)
deltaG.human <-estimate.deltaG(human.df, primer.pool, groupsOI = c(1,2,4,5,6,7,8, seq(9,20)), iterations = 100)

head(deltaG.zf)

compare.zf.cele <- mutate(melt(deltaG.cele, variable.name='group',value.name = 'dg'), species='cele') %>%
  bind_rows(., mutate(melt(deltaG.human, variable.name='group',value.name = 'dg') , species='human')) %>%
  bind_rows(., mutate(melt(deltaG.zf, variable.name='group',value.name = 'dg'), species='zf')) 


compare.zf.cele %>%
  filter(group %in% sample(compare.zf.cele$group, 10)) %>%
  rename(template=group) %>%
  inner_join(., dplyr::select(hex.df, template, dG), by='template') %>%
  ggplot(., aes(template,-dg, fill=species)) +
  geom_boxplot() +
  geom_point(aes(y=dG), color='red', size=2, shape=8)
  
  # ggplot(., aes(group, dg, fill=species)) + geom_boxplot()

compare.zf.cele %>%
  filter(group %in% sample(compare.zf.cele$group, 1000)) %>%
  rename(template=group) %>%
  inner_join(., dplyr::select(hex.df, template, dG, pt), by='template') %>%
  filter(pt>200) %>%
  ggplot(., aes(dG, dg, group=dG)) + geom_boxplot(alpha=0.1)

```
```{r}
compare.zf.cele %>%
  filter(group %in% sample(compare.zf.cele$group, 1000)) %>%
  rename(template=group) %>%
  # filter(!is.na(dg)) %>%
  dcast(template ~ species, value.var = 'dg',fun.aggregate = mean) %>%
  inner_join(., dplyr::select(hex.df, template, dG), by='template') %>%
  ggplot(., aes(dG, cele)) +
    geom_point()
  
```
## Find deltaG values mismatching pairs

```{r}
fit.gamma <- summary(fitdist(as.numeric(table(primer.pool)), 'gamma'))
gamma.shape = fit.gamma$estimate['shape']
gamma.rate = fit.gamma$estimate['rate']
pool.size = length(primer.pool)

keqs <- compute.keqs(cele.eps, human.eps = human.eps, zf.eps, gamma.shape, gamma.rate, pool.size=length(primer.pool), n.iterations = 10, take.pairs = T)
```
plotting a few distributions
```{r}
keqs %>%
  filter(template %in% sample(unique(keqs$template), 15)) %>%
  filter(pt>200) %>%
  ggplot(., aes(template, single.keq, fill=species)) + geom_boxplot(alpha=0.4)

```
Comparison with system method 
```{r}
joined.keqs <- keqs %>%
  inner_join(., t.groups, by='template') %>%
  mutate(group=as.numeric(group)) %>%
  inner_join(., melt(keq.zf, variable.name='group', value.name = 'keq.system'), by='group') 

# joined.keqs %>%
#   group_by(template) %>%
#   summarise(single.keq=mean(single.keq), keq.system=mean(keq.system)) %>%
#   filter(single.keq>-0.001) %>%
#   ggplot(., aes(single.keq, keq.system)) + geom_point(alpha=0.4)

joined.keqs %>%
  dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  mutate(single.keq=-single.keq) %>%
  melt(id.vars=c('p.iter','template', 'dG', 'species'), variable.name='keq.method') %>%
  filter(as.character(template) %in% sample(unique(joined.keqs$template), 10)) %>%
  # filter(pt>200) %>%
  ggplot(., aes(template, log(value), fill=keq.method)) + geom_boxplot()

joined.keqs %>%
  dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  mutate(single.keq=-single.keq) %>%
  melt(id.vars=c('p.iter','template', 'dG', 'species'), variable.name='keq.method') %>%
  filter(as.character(template) %in% sample(unique(joined.keqs$template), 10)) %>%
  # filter(pt>200) %>%
  ggplot(., aes(template, (value), fill=keq.method)) + geom_boxplot()

```
Comparison with Mfold
```{r}
joined.keqs %>%
  # dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  filter(as.character(template) %in% sample(unique(keqs$template), 10)) %>%
  filter(pt>200) %>%
  ggplot(., aes(template, log(-single.keq))) + geom_boxplot() +
  geom_point(aes(y=dG), color='red')

keqs %>%
  # dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  filter(as.character(template) %in% sample(unique(keqs$template), 600)) %>%
  filter(pt>300) %>%
  ggplot(., aes(exp(dG/0.59), single.keq)) +
  geom_point(alpha=0.1) 

```
```{r}
keqs %>%
  # dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  # filter(as.character(template) %in% sample(unique(keqs$template), 600)) %>%
  filter(pt>300) %>%
  group_by(template) %>%
  summarise(dG=first(dG), mean.keq=mean(single.keq, na.rm=T), sd.keq=sd(single.keq, na.rm=T), pt=first(pt)) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
  # head()
  # ggplot(., aes(mean.keq)) + geom_histogram()
  # filter(mean.keq<0 & mean.keq>-7e-04) %>%
    ggplot(., aes(dG/0.59, log(-mean.keq), color=nuc)) +
  geom_point(alpha=0.5) + 
    geom_errorbar(aes(ymax=log(-mean.keq+sd.keq), ymin=log(-mean.keq-sd.keq)), alpha=0.5)
  # xlim(-12, -8)
ggsave('~/AvOwork/Keq_vs_mfold.pdf')

```
Oh my God, yes.

Where do the pairs fall?
```{r}
keqs %>%
  filter(pt>300) %>%
  group_by(template) %>%
  summarise(dG=first(dG), mean.keq=mean(single.keq, na.rm=T), sd.keq=sd(single.keq, na.rm=T), pt=first(pt)) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
  mutate(label=ifelse(template=='ATAAAA' | template=='TTTTAT', template, '')) %>%
  ggplot(., aes(dG/0.59, log(-mean.keq), color=label)) +
  geom_point(alpha=0.5)  
  # geom_text_repel()
    geom_errorbar(aes(ymax=log(-mean.keq+sd.keq), ymin=log(-mean.keq-sd.keq)))


```


```{r}
keqs %>%
  # filter(pt>300) %>%
  dplyr::select(template, single.keq, species) %>%
  dcast(template ~ species, value.var='single.keq', fun.aggregate = mean) %>%
  mutate(nuc=sapply(template, prevalent_nucleotide)) %>%
  # summarise(cele=max(cele, na.rm=T))
  ggplot(., aes(log((-cele)), log((-human)), color=nuc)) + geom_point(alpha=0.3)
```

With pairs
```{r}
keqs.pairs <- compute.keqs(cele.eps, human.eps = human.eps, zf.eps, gamma.shape, gamma.rate, pool.size, n.iterations = 50, take.pairs = T)

pairs.keq <- keqs.pairs %>%
# dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
# filter(as.character(template) %in% sample(unique(keqs$template), 600)) %>%
filter(pt>200) %>%
group_by(pair) %>%
summarise(dG=first(dG), template=first(template), mean.keq=mean(single.keq, na.rm=T), sd.keq=sd(single.keq, na.rm=T), pt=first(pt))

cele.df.wKeq <- cele.all.df %>%
  add.pairs.info() %>%
  inner_join(dplyr::select(pairs.keq, pair, mean.keq), by='pair') 

cele.df.wKeq %>%
  mutate(frac.abundance = abundance/sum(as.numeric(abundance))) %>%
  group_by(template) %>%
  mutate(pred.cov=frac.abundance*(sum(0.00024*-mean.keq)/(1+sum(0.00024*-mean.keq)))) %>%
  ggplot(., aes(log(t.usage/sum(t.usage)), log(pred.cov/sum(pred.cov)))) + geom_point(alpha=0.3) +
  geom_abline(slope=1, intercept = 0, col='red')

keqs.pairs %>%
  # dplyr::select(template, dG, single.keq,p.iter, species, keq.system)  %>%
  # filter(as.character(template) %in% sample(unique(keqs$template), 600)) %>%
  filter(pt>500) %>%
  group_by(pair, species) %>%
  summarise(dG=first(dG), template=first(template), mean.keq=mean(single.keq, na.rm=T), sd.keq=sd(single.keq, na.rm=T), pt=first(pt)) %>%
  dplyr::select(pair, species,mean.keq) %>%
  dcast(pair ~ species) %>%
  mutate(nuc=sapply(pair, prevalent_nucleotide)) %>%
  ggplot(., aes(log(-zfish), log(-cele), color=nuc)) + geom_point(alpha=0.5)
  # filter(pair=='AAATTT.AAATTT')

head(pairs.keq)  

```

```{r}
keq.zf.df2 <- dg.match %>%
  inner_join(., t.groups, by='template') %>%
  mutate(group=as.numeric(group)) %>%
  inner_join(.,melt(keq.zf, variable.name='group', value.name = 'keq.system'), by='group') %>%
  group_by(template) %>%
  summarise(mean.system=median(keq.system), keq=first(single.keq), mfold=first(dG), p=first(p)) 

keq.zf.df2 %>%
  ggplot(., aes(mean.system, log(-keq), label=template)) + geom_point(alpha=0.2) + geom_text(size=2) 
  # ylim(-0.025, 0.01)

```


